safety_limiter/src/safety_limiter.cpp

```diff
@@ -139,258 +232,221 @@ class SafetyLimiterNode
+      pub_twist_->publish(cmd_vel);
 
       cloud_accum_.reset(new pcl::PointCloud<pcl::PointXYZ>);
       has_cloud_ = false;
       r_lim_ = 0;
 
-      diag_updater_.force_update();
+      diag_updater_->force_update();
       return;
     }
 
-    ros::Time now = ros::Time::now();
+    auto now = this->now();
     const double r_lim_current = predict(twist_);
 
     if (r_lim_current < r_lim_)
       r_lim_ = r_lim_current;
 
     if (r_lim_current < 1.0)
-      hold_off_ = now + hold_;
-
-    cloud_clear_ = true;
-
-    diag_updater_.force_update();
-  }
-  void cbParameter(const SafetyLimiterConfig& config, const uint32_t /* level */)
-  {
-    boost::recursive_mutex::scoped_lock lock(parameter_server_mutex_);
-    hz_ = config.freq;
-    timeout_ = config.cloud_timeout;
-    disable_timeout_ = config.disable_timeout;
-    vel_[0] = config.lin_vel;
-    acc_[0] = config.lin_acc;
-    vel_[1] = config.ang_vel;
-    acc_[1] = config.ang_acc;
-    max_values_[0] = config.max_linear_vel;
-    max_values_[1] = config.max_angular_vel;
-    z_range_[0] = config.z_range_min;
-    z_range_[1] = config.z_range_max;
-    dt_ = config.dt;
-    d_margin_ = config.d_margin;
-    d_escape_ = config.d_escape;
-    yaw_margin_ = config.yaw_margin;
-    yaw_escape_ = config.yaw_escape;
-    downsample_grid_ = config.downsample_grid;
-    hold_ = ros::Duration(std::max(config.hold, 1.0 / hz_));
-    allow_empty_cloud_ = config.allow_empty_cloud;
-
-    tmax_ = 0.0;
-    for (int i = 0; i < 2; i++)
     {
-      auto t = vel_[i] / acc_[i];
-      if (tmax_ < t)
-        tmax_ = t;
+      hold_off_ = now + hold_;
     }
-    tmax_ *= 1.5;
-    tmax_ += std::max(d_margin_ / vel_[0], yaw_margin_ / vel_[1]);
-    r_lim_ = 1.0;
+    RCLCPP_DEBUG(get_logger(), "safety_limiter: r_lim = %0.3f, hold_off = %0.3f", r_lim_, hold_off_.seconds());
+    cloud_clear_ = true;
+    diag_updater_->force_update();
   }
-  double predict(const geometry_msgs::Twist& in)
+  double predict(const geometry_msgs::msg::Twist& in)
   {
     if (cloud_accum_->size() == 0)
     {
       if (allow_empty_cloud_)
       {
         return 1.0;
       }
-      ROS_WARN_THROTTLE(1.0, "safety_limiter: Empty pointcloud passed.");
+      RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 1000, "safety_limiter: Empty pointcloud passed.");
       return 0.0;
     }
 
-    const bool can_transform = tfbuf_.canTransform(
-        base_frame_id_, cloud_accum_->header.frame_id,
-        pcl_conversions::fromPCL(cloud_accum_->header.stamp));
-    const ros::Time stamp =
-        can_transform ? pcl_conversions::fromPCL(cloud_accum_->header.stamp) : ros::Time(0);
-
-    geometry_msgs::TransformStamped fixed_to_base;
+    const rclcpp::Time pcl_stamp = pcl_conversions::fromPCL(cloud_accum_->header.stamp);
+    const bool can_transform = tfbuf_.canTransform(base_frame_id_, cloud_accum_->header.frame_id, pcl_stamp);
+    const rclcpp::Time stamp = can_transform ? pcl_stamp : rclcpp::Time(0);
```

Comment:
Note that the `clock_type` of the instance created by `pcl_conversions::fromPCL()` is also RCL_SYSTEM_TIME. Here, the problem does not occur because `pcl_stamp` is only used as arguments for tf2 methods in which  `clock_type` is ignored.